bindings/js/api.js

```diff
@@ -0,0 +1,283 @@
+      .subarray(ptr, ptr + len)));
+    return new Uint8Array(result);
+  };
+
+  const callWithProto = function(proto, args, loc_name) {
+    var expected_args_number = proto.args.filter(arg => arg.interface_index !== undefined).length;
+    if (args.length != expected_args_number) {
+      throw Error("wrong number of arguments to call the F*-wasm function " + loc_name + ": expected " + expected_args_number + ", got " + args.length)
+    }
+    let memory = new Uint32Array(Module.Kremlin.mem.buffer);
+    let sp = memory[0];
+    var var_lengths = {};
+    // Populating the variable length arguments by retrieving buffer lengths
+    proto.args.map((arg) => {
+      if (arg.type === "buffer") {
+        if ((typeof arg.size === "string") && (arg.interface_index !== undefined)) {
+          let func_arg = args[arg.interface_index];
+          var_lengths[arg.size] = func_arg.length;
+        }
+      }
+      if ((arg.type === "int") && (arg.interface_index !== undefined)) {
+        let func_arg = args[arg.interface_index];
+        var_lengths[arg.name] = func_arg;
+      }
+    });
+    // Retrieving all input buffers and allocating them in the Wasm memory
+    let args_pointers = proto.args.map((arg, i) => {
+      if (arg.type === "buffer") {
+        var size;
+        if (typeof arg.size === "string") {
+          size = var_lengths[arg.size]
+        } else {
+          size = arg.size;
+        }
+        var argByteBuffer;
+        if (arg.kind === "input") {
+          let func_arg = args[arg.interface_index];
+          argByteBuffer = new Uint8Array(func_arg);
+        } else if (arg.kind === "output") {
+          argByteBuffer = new Uint8Array(size);
+        }
+        CheckIfByteArray(argByteBuffer, size, proto.name);
+        let pointer = copy_array_to_stack(argByteBuffer);
+        return {
+          "value": pointer,
+          "index": i
+        };
+      }
+      if (arg.type === "bool" || arg.type === "int") {
+        var value;
+        if (arg.interface_index === undefined) {
+          value = var_lengths[arg.name];
+        } else {
+          value = args[arg.interface_index];
+        }
+        return {
+          "value": value,
+          "index": i
+        };
+      }
+      throw Error("Unimplemented !");
+    });
+    // Calling the wasm function !
+    let call_return = Module[proto.module][proto.module + "_" + proto.name](
+      ...args_pointers.map(x => {
+        return x.value;
+      })
+    );
+    // Populating the JS buffers returned with their values read from Wasm memory
+    let return_buffers = args_pointers.filter(pointer =>
+      proto.args[pointer.index].kind === "output"
+    ).map(pointer => {
+      let protoRet = proto.args[pointer.index];
+      var size;
+      if (typeof protoRet.size === "string") {
+        size = var_lengths[protoRet.size];
+      } else {
+        size = protoRet.size;
+      }
+      let retBuf = new ArrayBuffer(size);
```

Comment:
why is this not just `return read_memory(pointer.value, size)`? seems like you're doing an extra copy

Reply:
Indeed, `read_memory` already allocates a new buffer (not just a view into the wasm memory), so this here was unnecessary belt and straps.